core/src/impl/Kokkos_HostSpace_deepcopy.cpp

```diff
@@ -0,0 +1,134 @@
+#include "Kokkos_HostSpace_deepcopy.hpp"
+
+namespace Kokkos {
+
+namespace Impl {
+
+/*
+ * TODO: change casting to whatever C++ likes (reinterpret_cast)
+ *       adjust logic with nthreads==1 to test if inside a parallel region
+ *       the reason for arranging the logic this way, is because
+ *       calling the omp_* API can incur some latency.
+ */
+
+void hostspace_parallel_deepcopy(void * dst, const void * src, size_t n) {
+  constexpr size_t page_sz_ = 4*1024;
+
+  // on KNL avoid having each thread requesting whole pages of memcpy
+  // on HSW, this may be equal to bytes_per_chunk
+  constexpr size_t bytes_per_chunk = page_sz_/2;
+
+  // make sure this is atleast 1 page, so that the calculation
+  // in the remainder below is valid unisgned arithmetic
+  if( n < 3*page_sz_)  {
+    std::memcpy( dst, src, n );
+    return;
+  }
+
+  // I can't use exec_space::concurrency here (or can I?)
+  // this may require specializing for each execSpace...
+  #ifdef _OPENMP
+    const int nthreads = omp_get_max_threads();
+  #else
+    const int nthreads = 1;
+  #endif
+
+  if(nthreads == 1) {
+    std::memcpy( dst, src, n );
+    return;
+  }
+
+  #pragma omp parallel
+  {
+    // otherwise, execute a parallel loop over memcpy,
+    // copying pages at a time
+  
+    // This line tests if dst is page aligned. We don't 
+    // care about src, since it is readonly.
+    // Effectively, mask out the lower bits that account for page_sz-1.
+    // if the resulting address is the same as dst, then dst is page 
+    // aligned to begin with we also are promised that dst + page_sz 
+    // will always be valid, because we enforce that 2*page_sz the smallest
+    // work allowed in this region of code.
+    const unsigned char * pg_aligned_start = ( (((uintptr_t) dst) & ~(page_sz_-1)) == ((uintptr_t) dst) )
+                                              ? (unsigned char *) dst
+                                              : (unsigned char *)(  (((uintptr_t) dst) & ~(page_sz_-1)) + page_sz_);
+    // handle the remainder. How many bytes are there from dst to pg_aligned_start
+    const size_t remainder = (size_t ) (pg_aligned_start - ((unsigned char *) dst) );
+    const unsigned char * corrected_src_from_pg_aligned_dst = (unsigned char *) (((uintptr_t) src) + remainder);
+    const size_t new_n =  (n-remainder);
+    const size_t chunks = (new_n / bytes_per_chunk ) + 1;
+  
+    #ifdef JJE_DEBUG_PAR_DEEPCOPY
```

Comment:
Rename to something like KOKKOS_DEBUG_PRINT_PAR_DEEPCOPY_DIAGNOSTIC
or so

Reply:
It isn't obvious this should hang around.  It was from my development. It makes assumptions about the data being copied (64 bit scalars). I'd prefer to remove it.

I have a separate test program that stresses this algorithm (using all possible alignments and possible starting points for the pointers). That test program takes a few minutes to run (it sweeps a 4KB ^2 problem space with verification at each step...) The latter is to say, I do have a unit test for this, but it is more comprehensive than a standard unit test. (it is also non-kokkos, it just allocates memory and passes them to my function, which expects void*)

TLDR: I'd like to delete the debug stuff, I don't think it belongs in a library.